File: Day 29 17-11-24/Leet Code/solution-1.py
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # Create a list 'dp' where dp[i] will hold the minimum number of coins needed to make amount 'i'
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0  # Base case: No coins are needed to make the amount 0

        # Iterate over each coin in the list of coins
        for coin in coins:
            # For each coin, update the dp list for all amounts from the coin's value up to the target amount
            for i in range(coin, amount + 1):
                # Update dp[i] to be the minimum of its current value and the value of dp[i - coin] + 1
                # This represents using one more coin of the current denomination
                dp[i] = min(dp[i], dp[i - coin] + 1)

        # If dp[amount] is still infinity, it means it's not possible to make the amount with the given coins
        # Return -1 in such a case, otherwise return dp[amount] which is the minimum number of coins needed
        return dp[amount] if dp[amount] != float('inf') else -1

# Example usage: Calculate the fewest number of coins needed to make up the amount
solution = Solution()
print(solution.coinChange([1, 2, 5], 11))  # Output: 3
print(solution.coinChange([2], 3))        # Output: -1
print(solution.coinChange([1], 0))        # Output: 0

File: Day 29 17-11-24/Leet Code/solution-1.js
// Problem: Coin Change
// You are given an integer array coins representing coins of different denominations and an integer amount representing a total amount of money.
// Return the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.
// You may assume that you have an infinite number of each kind of coin.
// Constraints:
// 1 <= coins.length <= 12
// 1 <= coins[i] <= 2^31 - 1
// 0 <= amount <= 10^4

const coinChange = (coins, amount) => {
    // Create an array to store the minimum coins needed for each amount
    const dp = new Array(amount + 1).fill(Infinity);
    dp[0] = 0; // Base case: 0 coins are needed to make the amount 0

    // Iterate over each coin
    for (let coin of coins) {
        // Update the dp array for each amount from coin to amount
        for (let i = coin; i <= amount; i++) {
            dp[i] = Math.min(dp[i], dp[i - coin] + 1);
        }
    }

    // If dp[amount] is still Infinity, it means the amount cannot be made up by any combination of the coins
    return dp[amount] === Infinity ? -1 : dp[amount];
}

// Example usage:
console.log(coinChange([1, 2, 5], 11)); // Output: 3
console.log(coinChange([2], 3)); // Output: -1
console.log(coinChange([1], 0)); // Output: 0
